chalk-solve/src/clauses.rs

```diff
@@ -355,6 +365,70 @@ fn program_clauses_that_could_match<I: Interner>(
     Ok(())
 }
 
+/// Generate `Implemented` clauses for a `dyn Trait` type. We need to generate
+/// `Implemented` clauses for all super traits, and for each trait we require
+/// its where clauses. (See #203.)
+fn push_dyn_ty_impl_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_ref: TraitRef<I>,
+    seen_traits: &mut FxHashSet<TraitId<I>>,
+) {
+    // We have some `dyn Trait`, and some `trait SuperTrait: WC`
+    // which is a super trait of `Trait` (including actually
+    // just being the same trait); then we want to push
+    // `Implemented(dyn Trait: SuperTrait) :- WC`.
+    let trait_datum = db.trait_datum(trait_ref.trait_id);
+    let wc = trait_datum
```

Comment:
Speaking of helper methods, maybe it'd be nice to have e.g.

```rust
trait_datum.where_clauses().substitute(...)
```

where `where_clauses` returns `Binders<&...>`  (basically encapsulating the `binders.map_ref(..)` call).

(I think this would be useful elsewhere, too.)